skbio/metadata/intersection.pyx

```diff
@@ -0,0 +1,513 @@
+"""
+Data structure for performing intersect queries on a set of intervals which
+preserves all information about the intervals (unlike bitset projection methods).
+
+:Authors: James Taylor (james@jamestaylor.org),
+          Ian Schenk (ian.schenck@gmail.com),
+          Brent Pedersen (bpederse@gmail.com)
+"""
+
+# Historical note:
+#    This module original contained an implementation based on sorted endpoints
+#    and a binary search, using an idea from Scott Schwartz and Piotr Berman.
+#    Later an interval tree implementation was implemented by Ian for Galaxy's
+#    join tool (see `bx.intervals.operations.quicksect.py`). This was then
+#    converted to Cython by Brent, who also added support for
+#    upstream/downstream/neighbor queries. This was modified by James to
+#    handle half-open intervals strictly, to maintain sort order, and to
+#    implement the same interface as the original Intersecter.
+
+#cython: cdivision=True
+
+import operator
+
+cdef extern from "stdlib.h":
+    int ceil(float f)
+    float log(float f)
+    int RAND_MAX
+    int rand()
+    int strlen(char *)
+    int iabs(int)
+
+cdef inline int imax2(int a, int b):
+    if b > a: return b
+    return a
+
+cdef inline int imax3(int a, int b, int c):
+    if b > a:
+        if c > b:
+            return c
+        return b
+    if a > c:
+        return a
+    return c
+
+cdef inline int imin3(int a, int b, int c):
+    if b < a:
+        if c < b:
+            return c
+        return b
+    if a < c:
+        return a
+    return c
+
+cdef inline int imin2(int a, int b):
+    if b < a: return b
+    return a
+
+cdef float nlog = -1.0 / log(0.5)
+
+cdef class IntervalNode:
+    """
+    A single node of an `IntervalTree`.
+
+    NOTE: Unless you really know what you are doing, you probably should us
+          `IntervalTree` rather than using this directly.
+    """
+    cdef float priority
+    cdef public object interval
+    cdef public int start, end
+    cdef int minend, maxend, minstart
+    cdef IntervalNode cleft, cright, croot
+
+    property left_node:
+        def __get__(self):
+            return self.cleft if self.cleft is not EmptyNode else None
+    property right_node:
+        def __get__(self):
+            return self.cright if self.cright is not EmptyNode else None
+    property root_node:
+        def __get__(self):
+            return self.croot if self.croot is not EmptyNode else None
+
+    def __repr__(self):
+        return "IntervalNode(%i, %i)" % (self.start, self.end)
+
+    def __cinit__(IntervalNode self, int start, int end, object interval):
+        # Python lacks the binomial distribution, so we convert a
+        # uniform into a binomial because it naturally scales with
+        # tree size.  Also, python's uniform is perfect since the
+        # upper limit is not inclusive, which gives us undefined here.
+        self.priority = ceil(nlog * log(-1.0/(1.0 * rand()/RAND_MAX - 1)))
```

Comment:
I echo @wasade's concern over `rand` and would question why the priority is random. Also, using casts instead of multiplying by one would be preferable. (I'm not sure how smart the Cython compiler is, but explicit casting is clearer anyways).